################
References
################

Here, you will find a list of relevant references. For a comprehensive list, please also consult bibliography of
`our paper <https://doi.org/10.1101/2022.04.26.489549>`_ and our suplementary materials.

-----------------------

************
Our Paper
************

If you use this package, citing our paper is highly appreciated!

.. code-block:: text

    Wang, K., Yang, Y., Wu, F., Song, B., Wang, X., & Wang, T. (2022). A Comparison of Dimension Reduction Methods for Cytometry by Time-of-Flight Data. bioRxiv.


----------------------


********************
Selected References
********************

- Stark, S. G. et al. SCIM: universal single-cell matching with unpaired feature sets. Bioinformatics 36, i919-i927 (2020).
- Ha, M. K. et al. Mass Cytometry and Single-Cell RNA-seq Profiling of the Heterogeneity in Human Peripheral Blood Mononuclear Cells Interacting with Silver Nanoparticles. Small 16, e1907674 (2020).
- Zhang, F. et al. Defining inflammatory cell states in rheumatoid arthritis joint synovial tissues by integrating single-cell transcriptomics and mass cytometry. Nat. Immunol. 20, 928-942 (2019).
- van der Maaten, L. & Hinton, G. Visualizing Data using t-SNE. Journal of Machine Learning Research 9, 2579-2605 (2008).
- McInnes, L., Healy, J. & Melville, J. UMAP: Uniform Manifold Approximation and Projection for Dimension Reduction. arXiv (2018) doi:10.48550/arxiv.1802.03426.
- Amodio, M. et al. Exploring single-cell data with deep multitasking neural networks. Nat. Methods 16, 1139-1145 (2019).
- Pierson, E. & Yau, C. ZIFA: Dimensionality reduction for zero-inflated single-cell gene expression analysis. Genome Biol. 16, 241 (2015).
- Moon, K. R. et al. Visualizing structure and transitions in high-dimensional biological data. Nat. Biotechnol. 37, 1482-1492 (2019).
- Ding, J., Condon, A. & Shah, S. P. Interpretable dimensionality reduction of single cell transcriptome data with deep generative models. Nat. Commun. 9, 2002 (2018).
- Haghverdi, L., Buettner, F. & Theis, F. J. Diffusion maps for high-dimensional single-cell analysis of differentiation data. Bioinformatics 31, 2989-2998 (2015).
- Chang, Q. et al. Imaging Mass Cytometry. Cytometry A 91, 160-169 (2017).
- Abe, K., Minoura, K., Maeda, Y., Nishikawa, H. & Shimamura, T. Model-based clustering for flow and mass cytometry data with clinical information. BMC Bioinformatics 21, 393 (2020).
- Kobak, D. & Linderman, G. C. Initialization is critical for preserving global data structure in both t-SNE and UMAP. Nat. Biotechnol. 39, 156-157 (2021).
- Becht, E. et al. Dimensionality reduction for visualizing single-cell data using UMAP. Nat. Biotechnol. 37, 38-44 (2018).
- Oetjen, K. A. et al. Human bone marrow assessment by single-cell RNA sequencing, mass cytometry, and flow cytometry. JCI Insight 3, (2018).
- van der Maaten, L. Accelerating t-SNE using Tree-Based Algorithms. Journal of Machine Learning Research 15, 3221-3245 (2014).
- Linderman, G. C., Rachh, M., Hoskins, J. G., Steinerberger, S. & Kluger, Y. Fast interpolation-based t-SNE for improved visualization of single-cell RNA-seq data. Nat. Methods 16, 243-245 (2019).
- Poličar, P. G., Stražar, M. & Zupan, B. openTSNE: a modular Python library for t-SNE dimensionality reduction and embedding. BioRxiv (2019) doi:10.1101/731877.
- Saelens, W., Cannoodt, R., Todorov, H. & Saeys, Y. A comparison of single-cell trajectory inference methods. Nat. Biotechnol. 37, 547-554 (2019).
- Kobak, D. & Berens, P. The art of using t-SNE for single-cell transcriptomics. Nat. Commun. 10, 5416 (2019).
- Van Gassen, S. et al. FlowSOM: Using self-organizing maps for visualization and interpretation of cytometry data. Cytometry A 87, 636-645 (2015).
- Weber, L. M. & Robinson, M. D. Comparison of clustering methods for high-dimensional single-cell flow and mass cytometry data. Cytometry A 89, 1084-1096 (2016).
- Satija, R., Farrell, J. A., Gennert, D., Schier, A. F. & Regev, A. Spatial reconstruction of single-cell gene expression data. Nat. Biotechnol. 33, 495-502 (2015).
- Kaushik, A. et al. CyAnno: A semi-automated approach for cell type annotation of mass cytometry datasets. Bioinformatics (2021) doi:10.1093/bioinformatics/btab409.
- Hawley, D. et al. RNA-Seq and CyTOF immuno-profiling of regenerating lacrimal glands identifies a novel subset of cells expressing muscle-related proteins. PLoS ONE 12, e0179385 (2017).
- Jackson, H. W. et al. The single-cell pathology landscape of breast cancer. Nature 578, 615-620 (2020).
- Levine, J. H. et al. Data-Driven Phenotypic Dissection of AML Reveals Progenitor-like Cells that Correlate with Prognosis. Cell 162, 184-197 (2015).
- Samusik, N., Good, Z., Spitzer, M. H., Davis, K. L. & Nolan, G. P. Automated mapping of phenotype space with single-cell data. Nat. Methods 13, 493-496 (2016).
- Mosmann, T. R. et al. SWIFT-scalable clustering for automated identification of rare cell populations in large, high-dimensional flow cytometry datasets, part 2: biological evaluation. Cytometry A 85, 422-433 (2014).
- Heiser, C. N. & Lau, K. S. A Quantitative Framework for Evaluating Single-Cell Data Structure Preservation by Dimensionality Reduction Techniques. Cell Rep. 31, 107576 (2020).
- Levina, E. & Bickel, P. The Earth Mover's distance is the Mallows distance: some insights from statistics. in Proceedings Eighth IEEE International Conference on Computer Vision. ICCV 2001 251-256 (IEEE Comput. Soc, 2001). doi:10.1109/ICCV.2001.937632.
- Konstorum, A., Vidal, E., Jekel, N. & Laubenbacher, R. Comparative analysis of linear and nonlinear dimension reduction techniques on mass cytometry data. BioRxiv (2018) doi:10.1101/273862.
- Davies, D. L. & Bouldin, D. W. A cluster separation measure. IEEE Trans. Pattern Anal. Mach. Intell. 1, 224-227 (1979).
- Zang, C. et al. High-dimensional genomic data bias correction and data integration using MANCIE. Nat. Commun. 7, 11305 (2016).
- Danon, L., Díaz-Guilera, A., Duch, J. & Arenas, A. Comparing community structure identification. J. Stat. Mech. 2005, P09008-P09008 (2005).
